File: src/main/java/com/liyongquan/linklist/DeleteDuplicates2.java
package com.liyongquan.linklist;

/**
 * 82. 删除排序链表中的重复元素 II
 * <p>
 * 给定一个排序链表，删除所有含有重复数字的节点，只保留原始链表中 没有重复出现 的数字。
 * <p>
 * 示例 1:
 * <p>
 * 输入: 1->2->3->3->4->4->5
 * 输出: 1->2->5
 * 示例 2:
 * <p>
 * 输入: 1->1->1->2->3
 * 输出: 2->3
 * <p>
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/remove-duplicates-from-sorted-list-ii
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */
public class DeleteDuplicates2 {
    /**
     * 两个指针，分别指向要删除起始位置和结束位置，然后再进行删除操作
     *
     * @param head
     * @return
     */
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = dummy, cur = head;
        while (cur != null) {
            //删除后面重复的元素
            boolean del = false;
            while (cur.next != null && cur.next.val == cur.val) {
                //删除cur.next的结点
                ListNode next = cur.next;
                cur.next = next.next;
                next.next = null;
                del = true;
            }
            //如果重复，当前结点也需要删除
            if (del) {
                pre.next = cur.next;
            } else {
                pre = cur;
            }
            cur = pre.next;
        }
        return dummy.next;
    }
}

File: algorithm/golang/linkList/deleteDuplicates2.go
package linkList

// 给定一个已排序的链表的头 head ， 删除原始链表中所有重复数字的节点，只留下不同的数字 。返回 已排序的链表 。
//
//
//
//示例 1：
//
//
//输入：head = [1,2,3,3,4,4,5]
//输出：[1,2,5]
//示例 2：
//
//
//输入：head = [1,1,1,2,3]
//输出：[2,3]
//
//
//提示：
//
//链表中节点数目在范围 [0, 300] 内
//-100 <= Node.val <= 100
//题目数据保证链表已经按升序 排列

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// 双指针
func deleteDuplicates2(head *ListNode) *ListNode {
	dummy := &ListNode{-101, nil}
	dummy.Next = head
	pre := dummy
	p1 := pre.Next
	for p1 != nil {
		p2 := p1
		for p2.Next != nil && p1.Val == p2.Next.Val {
			p2 = p2.Next
		}
		if p1 != p2 {
			// 需要删除节点
			pre.Next = p2.Next
			p2.Next = nil
			p1 = pre.Next
		} else {
			pre = p1
			p1 = p1.Next
		}
	}
	return dummy.Next
}
